File: Offer 35. 复杂链表的复制.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

// DFS
class Solution {
    Map<Node, Node> visited;

    public Node copyRandomList(Node head) {
        visited = new HashMap<>();
        return dfs(head);
    }

    private Node dfs(Node head) {
        if (head == null) {
            return null;
        }

        if (!visited.containsKey(head)) {
            Node copy = new Node(head.val);
            visited.put(head, copy);
            copy.next = dfs(head.next);
            copy.random = dfs(head.random);
            return copy;
        } else {
            return visited.get(head);
        }
    }
}

// BFS
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        
        // 创建哈希表记录访问情况
        Map<Node, Node> map = new HashMap<>();
        Node clone = new Node(head.val);
        map.put(head, clone);

        Queue<Node> queue = new LinkedList<>();
        queue.offer(head);

        while (!queue.isEmpty()) {
            Node node = queue.poll();
            if (node.next != null && !map.containsKey(node.next)) {
                map.put(node.next, new Node(node.next.val));
                queue.offer(node.next);
            }
            if (node.random != null && !map.containsKey(node.random)) {
                map.put(node.random, new Node(node.random.val));
                queue.offer(node.random);
            }
            // 连接节点
            map.get(node).next = map.get(node.next);
            map.get(node).random = map.get(node.random);
        }

        return clone;
    }
}

// 优化迭代
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }

        Node p = head;
        // 把所有next连接起来
        while (p != null) {
            Node clone = new Node(p.val);
            clone.next = p.next;
            p.next = clone;

            p = p.next.next;
        }

        // 把所有random连接起来
        p = head;
        while (p != null) {
            if (p.random == null) {
                p.next.random = null;
            } else {
                p.next.random = p.random.next;
            }
            p = p.next.next;
        }

        // 分离原链表和拷贝链表
        Node p1 = head, p2 = head.next, res = head.next;
        while (p1 != null) {
            p1.next = p1.next.next;
            if (p2.next != null) {
                p2.next = p2.next.next;
            }

            p1 = p1.next;
            p2 = p2.next;
        }

        return res;
    }
}

File: Offer 35. 复杂链表的复制.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

# 哈希
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return

        copyHead = Node(head.val)
        curr = head
        mp = {head: copyHead}
        while curr:
            copy = mp[curr]
            
            if curr.next:
                if curr.next in mp.keys():
                    copy.next = mp[curr.next]
                else:
                    mp[curr.next] = Node(curr.next.val)
                    copy.next = mp[curr.next]

            if curr.random:
                if curr.random in mp.keys():
                    copy.random = mp[curr.random]
                else:
                    mp[curr.random] = Node(curr.random.val)
                    copy.random = mp[curr.random]       

            curr = curr.next          

        return copyHead


# O(1)空间迭代
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return

        cur = head
        while cur:
            node = Node(cur.val, cur.next)
            cur.next = node
            cur = cur.next.next

        cur = head
        while cur:
            if cur.random:
                cur.next.random = cur.random.next
            cur = cur.next.next

        cur = res = head.next
        pre = head
        while cur.next:
            pre.next = pre.next.next
            cur.next = cur.next.next
            pre = pre.next
            cur = cur.next

        return res